### ISSUE TITLE
Best practice to ensure reproducibility

### ISSUE THREAD

**Issue by celsofranssa** [sdv-dev/SDV#2398]
I get different results every time I train a synthesizer. Is there any way to ensure reproducibility, such as using a seed?

**Comment by srinify** [comment]
Hi @celsofranssa at the moment, fitting a synthesizer isn't reproducible but sampling from a fitted synthesizer is. 

That is to say, you can reset the randomization seed to the sampling state right after fitting using the [`reset_sampling()` method](https://docs.sdv.dev/sdv/single-table-data/sampling/sample-realistic-data#reset_sampling) on the synthesizer.

```
synthesizer.fit(real_data)
w
synthetic_data_1 = synthesizer.sample(100)

synthesizer.reset_sampling()
synthetic_data_2 = synthesizer.sample(100)
```

The `synthetic_data_1` and `synthetic_data_2` DataFrame objects will be the same.

One workflow to incorporate reproducibility is to save the synthesizer as a `pkl` file (the sampling state will also be saved) at the desired state. Every time you load the `pkl` file and then run `sample()` the first time, the synthetic data will be the same. Note that calling `sample()` a second time will change its sampling state. Every synthesizer has it's own `save()` and `load()` functions to help (e.g. for GaussianCopulaSynthesizer here's the [save() method](https://docs.sdv.dev/sdv/single-table-data/modeling/synthesizers/gaussiancopulasynthesizer#save) and the [load() function](https://docs.sdv.dev/sdv/single-table-data/modeling/synthesizers/gaussiancopulasynthesizer#gaussiancopulasynthesizer.load) ).

Does this workflow provide the reproducibility you're looking for?

**Comment by celsofranssa** [comment]
It would be great if we could reproduce the training step, ensuring the same results regarding quality and diagnostic metrics.

**Comment by srinify** [comment]
@celsofranssa we have an existing feature request issue [here](https://github.com/sdv-dev/SDV/issues/2022) for reproducibility specifically for `fit()`. If you think that issue does a good job capturing your own needs, I can close this issue out so we can focus future discussions over there?

**Comment by daley-flips** [comment]
this has been working well for me

```
SEED = 42
np.random.seed(SEED)
torch.manual_seed(SEED)
```
lmk if this helps? also see:
https://github.com/sdv-dev/SDV/issues/299

**Comment by celsofranssa** [comment]
Maybe encapsulate it into the following function:

````python
import os
import random
import numpy as np
import torch

def seed_everything(seed=42):
    """
    Set the seed for various random number generators to ensure reproducibility.

    This function sets the seed for the following modules/libraries:
    - `random` (Python's built-in pseudo-random generator)
    - `numpy` (Numerical computations)
    - `torch` (PyTorch computations, both CPU and CUDA)
    - Environment variable `PYTHONHASHSEED` to control hash randomization

    Args:
        seed (int): The seed value to set for all random number generators. Default is 42.

    Notes:
        - Setting `PYTHONHASHSEED` ensures consistent hashing across runs.
        - For PyTorch:
            - `torch.manual_seed(seed)` sets the seed for CPU computations.
            - If CUDA is available, `torch.cuda.manual_seed(seed)` and `torch.cuda.manual_seed_all(seed)` set the seed for GPU computations.
            - Setting `torch.backends.cudnn.deterministic = True` ensures that CUDA selects deterministic algorithms, which can impact performance.
            - Setting `torch.backends.cudnn.benchmark = False` disables the inbuilt cudnn auto-tuner that selects the best convolution algorithm for your hardware, ensuring reproducibility.

    Example:
        >>> seed_everything(123)
        This will set all the relevant seeds to 123, ensuring reproducible results.

    References:
        - Python's `random` module: https://docs.python.org/3/library/random.html
        - NumPy's random seed: https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html
        - PyTorch reproducibility: https://pytorch.org/docs/stable/notes/randomness.html
    """
    os.environ['PYTHONHASHSEED'] = str(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed(seed)
        torch.cuda.manual_seed_all(seed)
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False

```
